File: pyfcstm/diagnostics/analyzers/transition_info.py
"""Informational transition diagnostics."""

from typing import TYPE_CHECKING, Iterable, List

from ...utils.validate import ModelDiagnostic

if TYPE_CHECKING:  # pragma: no cover
    from ..inspect import StateInfo, TransitionInfo


def collect_transition_infos(
    states: Iterable['StateInfo'],
    transitions: Iterable['TransitionInfo'],
) -> List[ModelDiagnostic]:
    """Collect non-blocking transition observations."""
    states_by_path = {state.path: state for state in states}
    diagnostics: List[ModelDiagnostic] = []
    diagnostics.extend(_composite_self_transition_infos(
        states_by_path,
        transitions,
    ))
    diagnostics.extend(_eventless_guardless_transition_infos(transitions))
    return diagnostics


def _composite_self_transition_infos(states_by_path, transitions) -> List[ModelDiagnostic]:
    diagnostics: List[ModelDiagnostic] = []
    for transition in transitions:
        if transition.from_path != transition.to_path:
            continue
        state = states_by_path.get(transition.from_path)
        if state is None or not state.is_composite:
            continue
        diagnostics.append(ModelDiagnostic(
            code='I_TRANSITION_TO_SELF_VIA_PARENT',
            severity='info',
            message=(
                f'Composite state {transition.from_path!r} transitions '
                'to itself and may intentionally re-enter children.'
            ),
            refs={
                'state_path': transition.from_path,
                'crosses_composite': True,
            },
        ))
    return diagnostics


def _eventless_guardless_transition_infos(transitions) -> List[ModelDiagnostic]:
    diagnostics: List[ModelDiagnostic] = []
    for transition in transitions:
        if transition.from_path == '[*]':
            continue
        if transition.event is not None or transition.guard is not None:
            continue
        diagnostics.append(ModelDiagnostic(
            code='I_TRANSITION_NEVER_EVENT_TRIGGERED',
            severity='info',
            message=(
                f'Transition {transition.from_path!r} -> '
                f'{transition.to_path!r} has no event or guard.'
            ),
            refs={
                'from_path': transition.from_path,
                'to_path': transition.to_path,
                'transition_span': None,
            },
        ))
    return diagnostics

File: editors/jsfcstm/src/diagnostics/analyzers/transition-info.ts
import type {ModelDiagnosticJson, StateInfo, TransitionInfo} from '../inspect';

export function collectTransitionInfos(
    states: StateInfo[],
    transitions: TransitionInfo[],
): ModelDiagnosticJson[] {
    const statesByPath = new Map(states.map(state => [state.path, state]));
    return [
        ...collectCompositeSelfTransitionInfos(statesByPath, transitions),
        ...collectEventlessGuardlessTransitionInfos(transitions),
    ];
}

function collectCompositeSelfTransitionInfos(
    statesByPath: Map<string, StateInfo>,
    transitions: TransitionInfo[],
): ModelDiagnosticJson[] {
    const out: ModelDiagnosticJson[] = [];
    for (const transition of transitions) {
        if (transition.from_path !== transition.to_path) continue;
        const state = statesByPath.get(transition.from_path);
        if (!state || !state.is_composite) continue;
        out.push({
            code: 'I_TRANSITION_TO_SELF_VIA_PARENT',
            severity: 'info',
            message: `Composite state ${JSON.stringify(transition.from_path)} transitions to itself and may intentionally re-enter children.`,
            span: null,
            refs: {
                state_path: transition.from_path,
                crosses_composite: true,
            },
        });
    }
    return out;
}

function collectEventlessGuardlessTransitionInfos(transitions: TransitionInfo[]): ModelDiagnosticJson[] {
    const out: ModelDiagnosticJson[] = [];
    for (let transitionIndex = 0; transitionIndex < transitions.length; transitionIndex += 1) {
        const transition = transitions[transitionIndex];
        if (transition.from_path === '[*]') continue;
        if (transition.event !== null || transition.guard !== null) continue;
        out.push({
            code: 'I_TRANSITION_NEVER_EVENT_TRIGGERED',
            severity: 'info',
            message: `Transition ${JSON.stringify(transition.from_path)} -> ${JSON.stringify(transition.to_path)} has no event or guard.`,
            span: null,
            refs: {
                from_path: transition.from_path,
                to_path: transition.to_path,
                transition_span: null,
                transition_index: transitionIndex,
            },
        });
    }
    return out;
}
